shell/platform/windows/flutter_host_window_controller.cc

```diff
@@ -0,0 +1,325 @@
+  return engine_;
+}
+
+void FlutterHostWindowController::DestroyAllWindows() {
+  if (!windows_.empty()) {
+    // Destroy windows in reverse order of creation.
+    for (auto it = std::prev(windows_.end());
+         it != std::prev(windows_.begin());) {
+      auto current = it--;
+      auto const& [view_id, window] = *current;
+      if (window->GetWindowHandle()) {
+        DestroyHostWindow(view_id);
+      }
+    }
+  }
+}
+
+WindowSize FlutterHostWindowController::GetWindowSize(
+    FlutterViewId view_id) const {
+  HWND const hwnd = windows_.at(view_id)->GetWindowHandle();
+  RECT frame_rect;
+  DwmGetWindowAttribute(hwnd, DWMWA_EXTENDED_FRAME_BOUNDS, &frame_rect,
+                        sizeof(frame_rect));
+
+  // Convert to logical coordinates.
+  auto const dpr = FlutterDesktopGetDpiForHWND(hwnd) /
+                   static_cast<double>(USER_DEFAULT_SCREEN_DPI);
+  frame_rect.left = static_cast<LONG>(frame_rect.left / dpr);
+  frame_rect.top = static_cast<LONG>(frame_rect.top / dpr);
+  frame_rect.right = static_cast<LONG>(frame_rect.right / dpr);
+  frame_rect.bottom = static_cast<LONG>(frame_rect.bottom / dpr);
+
+  auto const width = frame_rect.right - frame_rect.left;
+  auto const height = frame_rect.bottom - frame_rect.top;
+  return {static_cast<int>(width), static_cast<int>(height)};
+}
+
+void FlutterHostWindowController::HideWindowsSatellites(HWND opt_out_hwnd) {
+  if (disable_satellite_hiding_) {
+    return;
+  }
+
+  // Helper function to check whether |hwnd| is a descendant of |ancestor|.
+  auto const is_descendant_of = [](HWND hwnd, HWND ancestor) -> bool {
+    HWND current = ancestor;
+    while (current) {
+      current = GetWindow(current, GW_OWNER);
+      if (current == hwnd) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  // Helper function to check whether |window| owns a dialog.
+  auto const has_dialog = [](FlutterHostWindow* window) -> bool {
+    for (auto* const owned : window->GetOwnedWindows()) {
+      if (owned->GetArchetype() == WindowArchetype::dialog) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  for (auto const& [_, window] : windows_) {
+    if (window->GetWindowHandle() == opt_out_hwnd ||
+        is_descendant_of(window->GetWindowHandle(), opt_out_hwnd)) {
+      continue;
+    }
+
+    for (FlutterHostWindow* const owned : window->GetOwnedWindows()) {
+      if (owned->GetArchetype() != WindowArchetype::satellite) {
+        continue;
+      }
+      if (!has_dialog(owned)) {
+        ShowWindow(owned->GetWindowHandle(), SW_HIDE);
+      }
+    }
+  }
+}
+
+void FlutterHostWindowController::SendOnWindowChanged(
+    FlutterViewId view_id) const {
+  if (channel_) {
+    WindowSize const size = GetWindowSize(view_id);
+    channel_->InvokeMethod(
+        "onWindowChanged",
+        std::make_unique<EncodableValue>(EncodableMap{
+            {EncodableValue("viewId"), EncodableValue(view_id)},
```

Comment:
Not sure about the Window code but most other Flutter code uses constants for the key names, i.e. kViewIdKey.